src/Tree.hs

```diff
@@ -1,24 +1,33 @@
 {-# LANGUAGE FlexibleInstances #-}
-module Tree (pretty, pretty', Tree(Leaf, Branch)) where
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE DeriveFoldable #-}
+{-# LANGUAGE DeriveFunctor #-}
+module Tree (pretty, pretty', Tree(Leaf, Branch), calculateDepth) where
 
-import Data.List.NonEmpty ( toList, NonEmpty )
+import Data.List.NonEmpty
 import Data.Text (Text(), pack)
 import Data.Foldable (fold)
 
-import Data.List(intercalate)
 
 data Tree a = Leaf a
    | Branch a (NonEmpty (Tree a))
-  deriving (Show, Eq)
+  deriving (Show, Eq, Functor, Foldable)
+
+newtype Depth = Depth { depthValue :: Int } deriving (Eq, Show)
+  deriving Num via Int
+
+calculateDepth :: Tree a -> Tree (a, Depth)
+calculateDepth = aux 0 where
+ aux n (Leaf a) = Leaf (a, n)
+ aux n (Branch a as) = Branch (a, n) (fmap (aux (n+1)) as)
 
 pretty :: CustomShow a => Tree a -> String
-pretty = aux 0 where 
- aux n (Leaf x) = customShowN n x
- aux n (Branch x y) = intercalate "\n" (customShowN n x :
-                                           toList (fmap (aux (n + 1)) y))
+pretty =  foldMap nodeToString . calculateDepth where
```

Comment:
How about the following?:
`foldMap (("\n" ++) . nodeToString) . calculateDepth`

Reply:
Neat